### PR TITLE
[compiler] Optimize emission in normal (non-value) blocks

### PR DESCRIPTION
**Pull request by josephsavona**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #29998
* __->__ #29883
* #29882

In #29863 I tried to find a clean way to share code for emitting instructions between value blocks and regular blocks. The catch is that value blocks have special meaning for their final instruction — that's the value of the block — so reordering can't change the last instruction. However, in finding a clean way to share code for these two categories of code, i also inadvertently reduced the effectiveness of the optimization.

This PR updates to use different strategies for these two kinds of blocks: value blocks use the code from #29863 where we first emit all non-reorderable instructions in their original order, _then_ try to emit reorderable values. The reason this is suboptimal, though, is that we want to move instructions closer to their dependencies so that they can invalidate (merge) together. Emitting the reorderable values last prevents this.

So for normal blocks, we now emit terminal operands first. This will invariably cause _some_ of the non-reorderable instructions to be emitted, but it will intersperse reoderable instructions in between, right after their dependencies. This maximizes our ability to merge scopes.

I think the complexity cost of two strategies is worth the benefit, though i still need to double-check all the output changes.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/fbt/fbt-whitespace-within-text.expect.md [hunk 1]
```diff
@@ -31,15 +31,16 @@ import fbt from "fbt";
 const _ = fbt;
 function Component(t0) {
   const $ = _c(2);
-  const { value } = t0;
   let t1;
-  if ($[0] !== value) {
+  if ($[0] !== t0) {
+    const { value } = t0;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/fbt/fbt-whitespace-within-text.expect.md:36]: probably an artifact of MemoizeFbtOperandsInSameScope, need to double-check this

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/fbt/fbt-whitespace-around-param-value.expect.md [hunk 2]
```diff
@@ -29,15 +29,16 @@ import fbt from "fbt";
 const _ = fbt;
 function Component(t0) {
   const $ = _c(2);
-  const { value } = t0;
   let t1;
-  if ($[0] !== value) {
+  if ($[0] !== t0) {
+    const { value } = t0;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/fbt/fbt-whitespace-around-param-value.expect.md:34]: same, looks related to MemoizeFbtOperandsInSameScope

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/fbt/fbtparam-with-jsx-element-content.expect.md [hunk 3]
```diff
@@ -29,10 +29,11 @@ import { c as _c } from "react/compiler-runtime";
 import fbt from "fbt";
 
 function Component(t0) {
-  const $ = _c(4);
-  const { name, data, icon } = t0;
+  const $ = _c(2);
   let t1;
-  if ($[0] !== name || $[1] !== icon || $[2] !== data) {
+  if ($[0] !== t0) {
+    const { name, data, icon } = t0;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/fbt/fbtparam-with-jsx-element-content.expect.md:35]: likely also MemoizeFbtOperandsInSameScope

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/inner-memo-value-not-promoted-to-outer-scope-dynamic.expect.md [hunk 4]
```diff
@@ -24,69 +24,50 @@ function Component(props) {
 ```javascript
 import { c as _c } from "react/compiler-runtime";
 function Component(props) {
-  const $ = _c(15);
+  const $ = _c(7);
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/inner-memo-value-not-promoted-to-outer-scope-dynamic.expect.md:27]: hello

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/merge-scopes-callback.expect.md [hunk 5]
```diff
@@ -27,7 +27,7 @@ import { c as _c } from "react/compiler-runtime"; // @enableInstructionReorderin
 import { useState } from "react";
 
 function Component() {
-  const $ = _c(6);
+  const $ = _c(4);
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/merge-scopes-callback.expect.md:30]: nice

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/merged-scopes-are-valid-effect-deps.expect.md [hunk 6]
```diff
@@ -32,7 +32,7 @@ import { c as _c } from "react/compiler-runtime"; // @validateMemoizedEffectDepe
 import { useEffect } from "react";
 
 function Component(props) {
-  const $ = _c(5);
+  const $ = _c(6);
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/merged-scopes-are-valid-effect-deps.expect.md:35]: this regresses, need to investigate

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/repro-duplicate-instruction-from-merge-consecutive-scopes.expect.md [hunk 7]
```diff
@@ -29,30 +29,39 @@ import { c as _c } from "react/compiler-runtime";
 import { Stringify } from "shared-runtime";
 
 function Component(t0) {
-  const $ = _c(3);
+  const $ = _c(5);
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/repro-duplicate-instruction-from-merge-consecutive-scopes.expect.md:32]: also regresses

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/InstructionReordering.ts [hunk 8]
```diff
@@ -319,8 +379,7 @@ function getDepth(env: Environment, nodes: Nodes, id: IdentifierId): number {
     return node.depth;
   }
   node.depth = 0; // in case of cycles
-  let depth =
-    node.instruction != null && mayAllocate(env, node.instruction) ? 1 : 0;
+  let depth = node.reorderability === Reorderability.Reorderable ? 0 : 1;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Optimization/InstructionReordering.ts:382]: in earlier versions the depth calculation was wrong, we don't want to align to mayAllocate() but rather count non-reorderable instruction as having higher priority than reorderable ones.

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/meta-isms/repro-cx-assigned-to-temporary.expect.md [hunk 9]
```diff
@@ -53,22 +53,27 @@ import { identity } from "shared-runtime";
...
 function Component() {
-  const $ = _c(2);
+  const $ = _c(4);
   const theme = useTheme();
 
-  const t0 = cx({
-    "styles/light": true,
-    "styles/dark": theme.getTheme() === DARK,
-  });
+  const t0 = theme.getTheme();
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/meta-isms/repro-cx-assigned-to-temporary.expect.md:59]: i need to debug why the calls are getting reordered here, i suspect that the cx() call is simply getting inlined but i'm not sure

**Comment by josephsavona** [comment]
The example from shows that we still have room to improve though:

```js
const formatter = new Intl.NumberFormat();
function Component() {
  const [count, setCount] = useState(0);
  return (
    <div>
      <button onClick={() => setCount(count - 1)}>-1</button>
      <span>{formatter.format(count)}</span>
      <button onClick={() => setCount(count + 1)}>+1</button>
    </div>
  )
}
```

With this PR it gets 13 memo slots, on main its 11. But really all of the JSX is dependent on `count` and can be merged into a single slot. In this example the instruction reordering can only help so much because of the limitations of MergeScopesThatInvalidateTogether. That pass needs to understand transitive dependencies in order to optimize cases like this one.

File: compiler/packages/babel-plugin-react-compiler/src/Optimization/InstructionReordering.ts [hunk 10]
```diff
@@ -243,67 +245,125 @@ function reorderBlock(
...
+     * but we'll be able to intersperse the depending reorderable instructions in between
+     * them in a way that works better with scope merging.
+     */
+    for (const operand of eachTerminalOperand(block.terminal)) {
+      DEBUG && console.log(`(terminal operand)`);
+      DEBUG && print(env, locals, shared, seen, operand.identifier.id);
+      emit(env, locals, shared, nextInstructions, operand.identifier.id);
+    }
+    // Anything not emitted yet is globally reorderable
+    for (const id of Array.from(locals.keys()).reverse()) {
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/Optimization/InstructionReordering.ts:353]: Makes sense, emitting reversed instructions ensures that nodes and their dependencies are grouped together.

I'm guessing that it's almost never the case that two instructions share the same dependency as we serialize reads from the same named variables. e.g.
```js
// Here, not only are the StoreLocals of `x` and `y` non-reorderable, but also so
// are the `input` references (despite being reads of inferred-frozen values)
function useFoo(input) {
  const x = foo(input);
  const y = bar(input);
  return {x, y};
}
```

**Review by mofeiZ — APPROVED** [review]

**Comment by josephsavona** [comment]
oops, accidentally rebased this into the previous PR

**Pull request closed without merging**